File: scripts/frontend/audit-frontend-validations.js
const fs = require('fs');
const path = require('path');

const FRONTEND_DIR = path.join(__dirname, '../../frontend/src');
const BACKEND_DIR = path.join(__dirname, '../../backend/src/main/java/sgc');
const REPORT_FILE = path.join(__dirname, '../../frontend-backend-validation-comparison.md');

function getAllFiles(dirPath, ext, arrayOfFiles) {
    const files = fs.readdirSync(dirPath);
    arrayOfFiles = arrayOfFiles || [];

    files.forEach(function(file) {
        if (fs.statSync(dirPath + "/" + file).isDirectory()) {
            arrayOfFiles = getAllFiles(dirPath + "/" + file, ext, arrayOfFiles);
        } else {
            if (file.endsWith(ext)) {
                arrayOfFiles.push(path.join(dirPath, "/", file));
            }
        }
    });

    return arrayOfFiles;
}

function scanFrontend() {
    const results = [];
    const files = getAllFiles(FRONTEND_DIR, '.vue');

    files.forEach(file => {
        try {
            const content = fs.readFileSync(file, 'utf-8');
            const lines = content.split(/\r?\n/);

            lines.forEach((line, i) => {
                if (line.includes(':disabled')) {
                    results.push({
                        file: file,
                        line: i + 1,
                        type: 'disabled_check',
                        content: line.trim()
                    });
                }
                if (line.includes('required') && line.includes('<')) {
                    results.push({
                        file: file,
                        line: i + 1,
                        type: 'html_required',
                        content: line.trim()
                    });
                }
                if (line.includes(':rules') || line.includes('v-model')) {
                    const lower = line.toLowerCase();
                    if (lower.includes('rule') || lower.includes('valid')) {
                        results.push({
                            file: file,
                            line: i + 1,
                            type: 'vue_rule',
                            content: line.trim()
                        });
                    }
                }
            });
        } catch (e) {
            console.error(e);
        }
    });
    return results;
}

function scanBackend() {
    const results = [];
    const files = getAllFiles(BACKEND_DIR, '.java');

    files.forEach(file => {
        // Filter only DTOs/Requests
        if (!file.includes('Dto') && !file.includes('Request')) return;

        try {
            const content = fs.readFileSync(file, 'utf-8');
            const lines = content.split(/\r?\n/);

            lines.forEach((line, i) => {
                if (['@NotNull', '@NotEmpty', '@NotBlank', '@Size'].some(ann => line.includes(ann))) {
                    results.push({
                        file: file,
                        line: i + 1,
                        type: 'bean_validation',
                        content: line.trim()
                    });
                }
            });
        } catch (e) {
            console.error(e);
        }
    });
    return results;
}

function generateReport(feData, beData) {
    let md = "# Frontend vs Backend Validation Audit\n\n";

    md += "## Backend Validations (DTOs)\n";
    md += "| File | Line | Annotation | Code |\n";
    md += "|------|------|------------|------|\n";
    
    beData.forEach(item => {
        const annotation = item.content.split('(')[0].split(' ')[0];
        // Using concatenation to avoid backtick hell
        md += "| " + path.basename(item.file) + " | " + item.line + " | " + annotation + " | `" + item.content + "` |\n";
    });

    md += "\n## Frontend Validations (Vue)\n";
    md += "| File | Line | Type | Code |\n";
    md += "|------|------|------|------|\n";
    
    feData.forEach(item => {
        let content = item.content;
        if (content.length > 100) content = content.substring(0, 100) + '...';
        content = content.replace(/\|/g, '\\|'); // Escape pipe for markdown table
        md += "| " + path.basename(item.file) + " | " + item.line + " | " + item.type + " | `" + content + "` |\n";
    });

    fs.writeFileSync(REPORT_FILE, md, 'utf-8');
}

console.log("Scanning validations...");
const fe = scanFrontend();
const be = scanBackend();
generateReport(fe, be);
console.log(`Report generated: ${REPORT_FILE}`);

File: scripts/frontend/audit_frontend_validations.py
import os
import re

FRONTEND_DIR = 'frontend/src'
BACKEND_DIR = 'backend/src/main/java/sgc'
REPORT_FILE = 'frontend-backend-validation-comparison.md'

def scan_frontend():
    results = []
    for root, dirs, files in os.walk(FRONTEND_DIR):
        for file in files:
            if file.endswith('.vue'):
                path = os.path.join(root, file)
                try:
                    with open(path, 'r', encoding='utf-8') as f:
                        lines = f.readlines()
                except:
                    continue

                for i, line in enumerate(lines):
                    # Check for :disabled logic (often used for validation)
                    if ':disabled' in line:
                        results.append({
                            'file': path,
                            'line': i + 1,
                            'type': 'disabled_check',
                            'content': line.strip()
                        })
                    # Check for required attribute
                    if 'required' in line and '<' in line:
                        results.append({
                            'file': path,
                            'line': i + 1,
                            'type': 'html_required',
                            'content': line.strip()
                        })
                    # Check for explicit validation rules (v-validate or rules props - generic)
                    if ':rules' in line or 'v-model' in line:
                         if 'rule' in line.lower() or 'valid' in line.lower():
                            results.append({
                                'file': path,
                                'line': i + 1,
                                'type': 'vue_rule',
                                'content': line.strip()
                            })
    return results

def scan_backend():
    results = []
    for root, dirs, files in os.walk(BACKEND_DIR):
        for file in files:
            if file.endswith('.java') and ('Dto' in file or 'Request' in file):
                path = os.path.join(root, file)
                try:
                    with open(path, 'r', encoding='utf-8') as f:
                        lines = f.readlines()
                except:
                    continue

                for i, line in enumerate(lines):
                    if '@NotNull' in line or '@NotEmpty' in line or '@NotBlank' in line or '@Size' in line:
                        results.append({
                            'file': path,
                            'line': i + 1,
                            'type': 'bean_validation',
                            'content': line.strip()
                        })
    return results

def generate_report(fe_data, be_data):
    with open(REPORT_FILE, 'w', encoding='utf-8') as f:
        f.write("# Frontend vs Backend Validation Audit\n\n")

        f.write("## Backend Validations (DTOs)\n")
        f.write("| File | Line | Annotation | Code |\n")
        f.write("|------|------|------------|------|\n")
        for item in be_data:
            annotation = item['content'].split('(')[0].split()[0]
            f.write(f"| {os.path.basename(item['file'])} | {item['line']} | {annotation} | `{item['content']}` |\n")

        f.write("\n## Frontend Validations (Vue)\n")
        f.write("| File | Line | Type | Code |\n")
        f.write("|------|------|------|------|\n")
        for item in fe_data:
            # truncate long lines
            content = item['content'][:100] + ('...' if len(item['content']) > 100 else '')
            content = content.replace("|", "\\|")
            f.write(f"| {os.path.basename(item['file'])} | {item['line']} | {item['type']} | `{content}` |\n")

if __name__ == '__main__':
    print("Scanning validations...")
    fe = scan_frontend()
    be = scan_backend()
    generate_report(fe, be)
    print(f"Report generated: {REPORT_FILE}")
